Introduce `getInvalidations` for IsolatedKind

antoniofrighetto opened the pull request.

antoniofrighetto committed:
Loader: associate container name to its role

`Container` names were mistakenly associated to the name
of their declaration. This has been fixed by associating
them to their `Role`.

antoniofrighetto committed:
Context: fix entry in `containsReadOnlyContainer`

`ReadOnlyContainers` is defined to be a `llvm::StringMap`
of a further `llvm::StringMap` of `std::unique_ptr`s of
`ContainerBase`. A logic issue has been addressed by
properly dereferincing the inner map.

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -50,6 +52,107 @@ using TaggedFK = TaggedFunctionKind;
+
+  const auto &CurrentModel = getModelFromContext(Ctx);
+  if (not Ctx.containsReadOnlyContainer(BinaryCrossRelationsRole))
+    return;
```
The `BinaryCrossRelationsRole` is only for refining the set of things you need to remove.
If you don't have it, you must invalidate *all* the targets. Here you're invalidating no target at all.

antoniofrighetto replied:
The [upcoming structure](https://github.com/revng/revng/pull/209/commits/33268127353b18637111e97de73726f5c3c20988) of the Pipes guarantees the `BinaryCrossRelationsRole` is created if there at least a function has been isolated. If `CrossRelations` container does not exist, there are no isolated functions at all, thus returning should be reasonable.

antoniofrighetto replied:
It should be OK now, we're invalidating all the targets (by specifying only the rank), sounds good?

aleclearmind replied:
See the other comment

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -50,6 +52,107 @@ using TaggedFK = TaggedFunctionKind;
+  if (not Ctx.containsReadOnlyContainer(BinaryCrossRelationsRole))
+    return;
+
+  static constexpr auto BCRR = BinaryCrossRelationsRole;
```
```suggestion
  static constexpr auto &BCRR = BinaryCrossRelationsRole;
```
You're copying it here,I think.

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -50,6 +52,107 @@ using TaggedFK = TaggedFunctionKind;
+}
+
+using IsolatedFK = IsolatedFunctionKind;
+void IsolatedFK::getInvalidations(const Context &Ctx,
+                                  TargetsList &ToRemove,
+                                  const GlobalTupleTreeDiff &Diff) const {
```
You're using `auto` way too much. Please change this habit now.

**Do not abuse `auto`**. Use `auto` only if:
    1. The line is very long and you can't figure any other way to make `check-conventions` happy.
    2. The type is obvious since it's expressedly
 stated on the same line (e.g., `auto *Lol = dyn_cast<CallInst>(...)`).
    3. The type is very ugly (typically, iterators).

antoniofrighetto replied:
I believe I have used `auto` appropriately so far. For instance, you can infer that the following line returns a ContainerType of `FileContainer` kind:
```cpp
const auto &Container = Ctx.getReadOnlyContainer<FileContainer>(BCRR);
```
Whereas, e.g., here:
```cpp
auto AttrMatcher = PM::create<Binary>("/Functions/*/Attributes").value();
```
It is necessary to make `check-conventions` happy. For the same, if the type returns a long `std::optional`, I needed to opt for `auto`. All the other cases are continuously being reviewed in order not to abuse `auto`.

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -50,6 +52,107 @@ using TaggedFK = TaggedFunctionKind;
+  yield::CrossRelations Relations;
+  YAMLInput >> Relations;
+
+  const auto *ModelDiff = Diff.getAs<Binary>();
+  if (not ModelDiff)
+    return;
```
@drblallo Doesn't the diff include which global is subject to the change? We can't decide how to at only based on type.

drblallo replied:
sent to antonio the fix

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -50,6 +52,107 @@ using TaggedFK = TaggedFunctionKind;
+    return;
+
+  const Binary &Model = *CurrentModel.get();
+  auto EntryMatcher = PM::create<Binary>("/Functions/*/Entry").value();
```
As we said multiple times, the key *cannot* change. You should never see this.
If someone changes the key, you will not see a change on `/Functions/0x1000/Entry` you will see the whole function (`/Functions/0x1000`) being removed and added again (`/Functions/0x2000`).
Did you to make such a change and see what `revng model diff` does?

antoniofrighetto replied:
I realize it, but the regex `"/Functions/*/Entry"` used to work correctly to match an addition and a removal until not a long ago. That said, since not correct, I had changed it. The approach used now should be correct.

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -50,6 +52,107 @@ using TaggedFK = TaggedFunctionKind;
+                                            "PreservedRegisters")
+                           .value();
+  auto AttrMatcher = PM::create<Binary>("/Functions/*/Attributes").value();
+  auto DynFuncMatcher = PM::create<Binary>("/ImportedDynamicFunctions/"
+                                           "*/OriginalName")
+                          .value();
```
Ibidem. This is a key, it will never match.

antoniofrighetto replied:
Per above.

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -50,6 +52,107 @@ using TaggedFK = TaggedFunctionKind;
+
+  const Binary &Model = *CurrentModel.get();
+  auto EntryMatcher = PM::create<Binary>("/Functions/*/Entry").value();
+  auto FSOMatcher = PM::create<Binary>("/Types/RawFunctionType-*/"
+                                       "FinalStackOffset")
+                      .value();
+  auto PresRegsMatcher = PM::create<Binary>("/Types/RawFunctionType-*/"
+                                            "PreservedRegisters")
+                           .value();
```
I'm afraid these won't do. We're not considering `CABIFunctionType`.

I'm starting to think that the best approach, at this point, is to ignore the diff and just do:

```c++
for (const model::Function &NewFunction : NewModel.Functions) {
  const auto &OldFunction = OldModel.Functions.at(NewFunction.Entry);
  Layout Old = makeLayout(OldFunction);
  Layout New = makeLayout(NewFunction);

  if (Old.FSO != New.FSO) {
    // ...
  }

}
```

In the future we could refine things reasoning on the diff to be more efficient.

antoniofrighetto replied:
I don't understand. `CABIFunctionType` does not come with `FinalStackOffset`, the user cannot even change it, hence, why would we ever want to pass to `Layout` and check `FinalStackOffset` for `CABIFunctionType` prototypes?

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -50,6 +52,107 @@ using TaggedFK = TaggedFunctionKind;
+                          .value();
+
+  for (const auto &Change : ModelDiff->Changes) {
+    if (Change.Path.empty())
+      continue;
```
What? Assert?

drblallo committed:
Allow access to `Global` in invalidation

antoniofrighetto committed:
Move `ProcessCallGraph` pipe within `Isolate`

`CrossRelations` container needs to be available
just as `FunctionIsolation` has executed.

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -49,18 +55,223 @@ TaggedFunctionKind::symbolToTarget(const llvm::Function &Symbol) const {
+  // If CrossRelations is not found, the most conservative approach is to
+  // invalidate all targets.
+  if (not Ctx.containsReadOnlyContainer(BinaryCrossRelationsRole)) {
+    ToRemove.emplace_back(Target(*this));
```
Does this create a `*` target?
`*` is going to disappear soon, insert all the functions by hand.

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -49,18 +55,223 @@ TaggedFunctionKind::symbolToTarget(const llvm::Function &Symbol) const {
+  // old model, and the new one (the one with the change).
+  for (const model::Function &NewFunction : Model->Functions) {
+    MetaAddress Entry = NewFunction.Entry;
+    const auto &OldFunction = OldModel->Functions.at(Entry);
```
The function might have been removed. You can't use `.at`.

antoniofrighetto replied:
Wait sorry, I still have the function in the old model, no? Otherwise, what's the purpose of checking `FSO` & stuff if I do not have the function?

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -49,18 +55,223 @@ TaggedFunctionKind::symbolToTarget(const llvm::Function &Symbol) const {
+      if (Change.New) {
+        // Invalidate all targets
+        for (const model::Function &Function : Model->Functions)
+          Targets.insert(Function.Entry.toString());
```
Postpone `.toString` for the final step in which you build the targets.

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -49,18 +55,223 @@ TaggedFunctionKind::symbolToTarget(const llvm::Function &Symbol) const {
+        auto Function = std::get<model::Function>(*Change.Old);
+        Targets.insert(Function.Entry.toString());
+        insertCallers(CrossRelations, Function.Entry, Targets);
+      }
```
```suggestion
      } else {
        revng_abort();
      }
```

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -49,18 +55,223 @@ TaggedFunctionKind::symbolToTarget(const llvm::Function &Symbol) const {
+    // A new prototype has been added or removed?
+    if (pathAsString<Binary>(Change.Path) == "/Types") {
+      if (Change.New) {
+        // Invalidate all targets
+        for (const model::Function &Function : Model->Functions)
+          Targets.insert(Function.Entry.toString());
```
Why should we invalidate all the functions when a new type is added?

antoniofrighetto replied:
Indeed, we do not have too. Removed that part and kept it for `ABIEnforced` Kind, if anything.

aleclearmind replied:
Uhm, no, I don't see why adding a new type should trigger an invalidation in general.

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -49,18 +55,223 @@ TaggedFunctionKind::symbolToTarget(const llvm::Function &Symbol) const {
+        for (const model::Function &Function : Model->Functions)
+          Targets.insert(Function.Entry.toString());
+      } else if (Change.Old) {
+        auto Prototype = std::get<model::TypePath>(*Change.Old);
+
+        for (const model::Function &Function : Model->Functions) {
+          if (Prototype == Function.Prototype) {
+            // Invalidate the function and its callers
+            Targets.insert(Function.Entry.toString());
+            insertCallers(CrossRelations, Function.Entry, Targets);
+          }
+        }
+
+        for (const model::DynamicFunction &Function : IDF) {
+          if (Prototype == Function.Prototype) {
+            // Invalidate callers of the model::DynamicFunction
+            insertCallers(CrossRelations, Function.OriginalName, Targets);
+          }
+        }
```
Given what we do below, I don't think this is necessary.
Even if the type is completely changed, as long as the FSO and set of preserved registers is the same, we don't need to invalidate.

antoniofrighetto replied:
Uhmm, right, good catch, thanks. I drop it here, and likely reuse it for `ABIEnforced` Kind.

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -49,18 +55,223 @@ TaggedFunctionKind::symbolToTarget(const llvm::Function &Symbol) const {
+      bool NewAttribute = std::get<FunctionAttribute>(*Change.New);
+      if (NewAttribute == FunctionAttribute::NoReturn) {
+        // Invalidate callers
+        insertCallers(CrossRelations, Entry, Targets);
```
Why not invalidate the function itself? Comment.

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -49,18 +55,223 @@ TaggedFunctionKind::symbolToTarget(const llvm::Function &Symbol) const {
+    if (OldFunction.Prototype != NewFunction.Prototype)
+      continue;
+
+    auto Old = abi::FunctionType::Layout::make(OldFunction.Prototype);
+    auto New = abi::FunctionType::Layout::make(NewFunction.Prototype);
+
+    if (Old.FinalStackOffset != New.FinalStackOffset
+        or Old.CalleeSavedRegisters != New.CalleeSavedRegisters) {
+      Targets.insert(Entry.toString());
+      insertCallers(CrossRelations, Entry, Targets);
+    }
```
Factor out common code with the part below.

antoniofrighetto replied:
Sounds good.

aleclearmind commented on lib/Pipes/TaggedFunctionKind.cpp:
```diff
@@ -49,18 +55,223 @@ TaggedFunctionKind::symbolToTarget(const llvm::Function &Symbol) const {
+      }
+    }
+
+    // A new dynamic function has been added or removed?
+    if (pathAsString<Binary>(Change.Path) == "/ImportedDynamicFunctions") {
+      if (Change.New) {
+        // Invalidate all targets
+        for (const model::Function &Function : Model->Functions)
+          Targets.insert(Function.Entry.toString());
+      } else if (Change.Old) {
+        // Invalidate callers of the model::DynamicFunction
+        for (const model::DynamicFunction &Function : IDF)
+          insertCallers(CrossRelations, Function.OriginalName, Targets);
+      }
+    }
+  }
```
I think we want to do this for regular functions too.
You probably got it mixed up with the types and functions above.

antoniofrighetto replied:
We invalidate all targets if we have an addition, or invalidate the function itself (not needed for if we have a `model::DynamicFunction` & its callers. This is what is happening currently both for `model::Function` and `model::DynamicFunction`. I didn't understand what is making you perplexed.

aleclearmind replied:
OK, put them close together, they are almost the same code.

antoniofrighetto replied:
Sure.

antoniofrighetto committed:
Pipeline: refine overestimated targets

A logic issue was addressed in `Runner::getDiffInvalidations`.
Premature invalidation of all targets (overestimated targets
found by `getInvalidations`) could lead to an error due to
lack of those targets (do not exist yet). We make sure that
such targets exist before providing them to the pipeline.

drblallo committed:
Introduce `ApplyDiffAnalysis`

antoniofrighetto committed:
Runner: bail out with no requests target

antoniofrighetto committed:
CLOption: improve `isSet` for a `cl::opt<T>`

A logic issue existed when validating whether a `cl::opt<T>`
was set. This has been addressed by replacing `isDefaultOption`
method with `getNumOccurrences`, as the former one was not
meant to be used for checking if the option was set or not.
Likewise, the default Option itself is written to a `std::string`,
in method `get`; the latter is leveraged by `Invokable`.

antoniofrighetto committed:
`revng.daemon`: accept no arguments for analysis

Let GraphQL accept zero or more parameters for analyses.

antoniofrighetto committed:
Introduce `print-invalidations` tool

A simple tool that prints the overestimated targets
found by `getInvalidations` of all Kinds, without
actually invalidating them.

antoniofrighetto committed:
Introduce `IsolatedKind::getInvalidations`

Targets may be no longer valid upon changes to the model done by
the user, thus they require to be invalidated. Targets associated
to `IsolatedFunctionKind`, whose Kind is produced by `IsolatePipe`
and handled in `IsolateFunctions`, are now recomputed via newly-
added `getInvalidations`.

antoniofrighetto committed:
Introduce test suite for invalidation work

`for-invalidation.yml` introduces the configuration
to check the diffs between the original model and
the patched one. Diffs is a set containing changes
to, e.g., `FinalStackOffset`, `PreservedRegisters`,
name of `Types`, and other changes to the model.

antoniofrighetto committed:
FunctionIsolation: execute `CollectCFG` earlier

This change makes sure that CollectCFG is always intertwined
with FunctionIsolation, as it needs to run before the latter
executed. To avoid possible rearrangements on the pipeline
in the future, CollectCFG pass is now registered within FI.

antoniofrighetto committed:
Yield: split internal pipes off

A cyclic dependency between shared objects `revngYield` and
`revngFunctionIsolation` could exist if `FunctionIsolation`
were to use `yield`-defined methods. This has been prevented
by splitting pipes off within `Yield`.

antoniofrighetto committed:
ProcessCallGraph: fix access to `Metadata`

A logic issue existed in the Pipe, as it was mistakenly relying on
`IsolatedRoot` instead of the isolated function. That was breaking
things if those functions were to be removed. This issue has been
addressed by correcting the behaviour of the Pipe.

antoniofrighetto committed:
TupleTreeDiff: evict cached references in `apply`

As the input model requires to be mutable, `evictCachedReferences`
is needed in order to invalidate cached references.

antoniofrighetto commented:
Requires [`revng-qa#11`](https://github.com/revng/revng-qa/pull/11).

aleclearmind commented:
NOTETOSELF: resume reviewing from "Pipeline: refine overestimated targets " included.

Cvelth commented on lib/Yield/CMakeLists.txt:
```diff
@@ -85,3 +81,17 @@ target_tuple_tree_generator(
+target_link_libraries(
+  revngYieldPipes
+  revngYield
+  revngFunctionIsolation
```
Yep, as far as I see now, `CrossRelations` only depends on `yield::Graph` (outside of the `model` and `efa` dependencies) so this should not be a problem.

My only concern is what I should rename the `yield::Graph` to then.

aleclearmind replied:
Its core property seems to a graph where nodes have a position in the plain. `LaidOutGraph`?

Cvelth replied:
I guess that would work, but I'm not excited. Since the graph does not get created in `LaidOff` state, you need to run a `Layouter` separately. This could lead to confusion.

Also, where's the boundary we're cutting stuff on? I feel like wherever the graph goes, the layouter should go too, but I guess that's it.

Should Graph + layouters + relations should be a single library? Or two distinct ones (as in, graph stuff has nothing to do with the relations)?

aleclearmind replied:
I feel like we should decouple the emission of graphs from the laying out.

Something like:
```c++
GenericGraph emitCrossRelations();
GenericGraph emitCFG();
yield::Graph layOut(GenericGraph);
```
Is there something preventing us to go this route?

Cvelth replied:
This would make the layout function super-specialized. But maybe that's not such a bad thing, tbh.

But I'd still prefer for both graphs to be explicitly named so it's clear what they are (of course they are both `GenericGraph`s).

aleclearmind replied:
Can you link the pad?

Cvelth replied:
I linked it in `~asm-and-graphs` header, but sure, here it is: https://pad.rev.ng/s/S3VedMpAo

aleclearmind commented on lib/Pipeline/Runner.cpp:
```diff
@@ -383,23 +392,39 @@ const KindsRegistry &Runner::getKindsRegistry() const {
+                                  const Global &After,
                                   InvalidationMap &Map) const {
+  TargetsList OverestimatedTargets;
+  for (const Kind &Rule : getKindsRegistry())
+    Rule.getInvalidations(getContext(),
+                          OverestimatedTargets,
+                          Diff,
+                          Before,
+                          After);
```
```suggestion
  for (const Kind &Kind : getKindsRegistry()) {
    Kind.getInvalidations(getContext(),
                          OverestimatedTargets,
                          Diff,
                          Before,
                          After);
  }
```

aleclearmind commented on include/revng/Pipeline/Option.h:
```diff
@@ -48,15 +48,27 @@ template<typename T, size_t Index>
-using OptionType = typename ArgTypeImpl<InvokableType, Index>::Type;
+using OptionTypeImpl = typename ArgTypeImpl<InvokableType, Index>::Type;
+
+// clang-format off
+template<typename InvokableType, size_t Index>
+constexpr bool IsConstCharPtr = std::is_same_v<OptionTypeImpl<InvokableType,
+                                                              Index>,
+                                               const char *>;
+
+template<typename InvokableType, size_t Index>
+using OptionType = std::conditional_t<IsConstCharPtr<InvokableType, Index>,
+                                      std::string,
+                                      OptionTypeImpl<InvokableType, Index>>;
+// clang-format on
```
Why `clang-format off`?

aleclearmind commented on include/revng/Pipeline/CLOption.h:
```diff
@@ -73,8 +73,14 @@ class CLOptionWrapper : public CLOptionBase {
-  bool isSet() const override { return not Option.isDefaultOption(); }
-  llvm::StringRef get() const override { return Option.ValueStr; }
+  bool isSet() const override { return Option.getNumOccurrences() != 0; }
+  std::string get() const override {
+    std::string ToReturn;
+    llvm::raw_string_ostream SO(ToReturn);
+    SO << Option.getValue();
+    SO.flush();
+    return ToReturn;
+  }
```
Commit message is too vague

aleclearmind replied:
The commit still does not address this part of the change.

aleclearmind commented on lib/Yield/CrossRelations.cpp:
```diff
@@ -143,19 +155,23 @@ yield::Graph yield::CrossRelations::toYieldGraph() const {
+                                                     ranks::Instruction);
+      auto ToKey = p::genericLocationFromString<0>(ToLocation,
                                                    ranks::Function,
                                                    ranks::BasicBlock,
                                                    ranks::Instruction);
-    auto ToKey = p::genericLocationFromString<0>(ToLocation,
-                                                 ranks::Function,
-                                                 ranks::BasicBlock,
-                                                 ranks::Instruction);
-    revng_assert(FromKey.has_value() && ToKey.has_value());
-
-    auto *FromNode = LookupHelper.at(std::get<0>(FromKey.value()));
-    FromNode->addSuccessor(LookupHelper.at(std::get<0>(ToKey.value())));
+      revng_assert(FromKey.has_value() && ToKey.has_value());
+
+      auto *FromNode = LookupHelper.at(std::get<0>(FromKey.value()));
+      auto *ToNode = LookupHelper.at(std::get<0>(ToKey.value()));
+      if (!llvm::is_contained(FromNode->successors(), ToNode))
+        FromNode->addSuccessor(ToNode);
+    }
```
Very vague commit message

antoniofrighetto replied:
Improved message with better description and split it into two different commits.

aleclearmind commented on lib/Yield/Pipes/CallGraphPipes.cpp:
```diff
@@ -35,7 +35,10 @@ void ProcessCallGraph::run(pipeline::Context &Context,
+
+  // If some functions are missing, do not output anything
+  if (Metadata.size() != Model->Functions.size())
+    return;
```
Commit message is very vague. If you got these commits from other people, please ask them to provide an accurate commit message.

aleclearmind commented on tools/pipeline/print-invalidations/Main.cpp:
```diff
@@ -0,0 +1,88 @@
+
+#include <system_error>
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/DynamicLibrary.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/ToolOutputFile.h"
+#include "llvm/Support/raw_os_ostream.h"
+
+#include "revng/Model/LoadModelPass.h"
+#include "revng/Pipeline/AllRegistries.h"
+#include "revng/Pipeline/ContainerSet.h"
+#include "revng/Pipeline/CopyPipe.h"
+#include "revng/Pipeline/GenericLLVMPipe.h"
+#include "revng/Pipeline/Global.h"
+#include "revng/Pipeline/LLVMContainerFactory.h"
+#include "revng/Pipeline/LLVMGlobalKindBase.h"
+#include "revng/Pipeline/Loader.h"
+#include "revng/Pipeline/Step.h"
+#include "revng/Pipeline/Target.h"
+#include "revng/Pipes/ModelGlobal.h"
+#include "revng/Pipes/PipelineManager.h"
+#include "revng/Pipes/ToolCLOptions.h"
+#include "revng/Support/CommandLine.h"
+#include "revng/Support/InitRevng.h"
+#include "revng/TupleTree/TupleTree.h"
```
I doubt we really need all of these includes. Please minimize the list.

aleclearmind commented on tools/pipeline/print-invalidations/Main.cpp:
```diff
@@ -0,0 +1,88 @@
+  cantFail(std::move(EL));
+
+  TargetsList OverestimatedTargets;
+  for (const Kind &Rule : Manager.getRunner().getKindsRegistry())
+    Rule.getInvalidations(Manager.context(),
+                          OverestimatedTargets,
+                          Diff,
+                          OriginalModel,
+                          *AfterModel);
```
Drop all references to "Rule". Leftover of old naming.

aleclearmind commented on tools/pipeline/print-invalidations/Main.cpp:
```diff
@@ -0,0 +1,88 @@
+  auto Diff = AbortOnError(deserializeFileOrSTDIN<DiffRef>(DiffPath));
+  revng::ErrorList EL;
+  Diff.apply(AfterModel->get(), EL);
+  cantFail(std::move(EL));
```
What's going on here?

antoniofrighetto replied:
`AbortOnError` was replaced with `cantFail`, after the introduction of `ErrorList`, and more specifically, after [this](https://github.com/revng/revng/commit/0b80801b2b75b3584b811148994a16dc4f184bf2) jackv's change, whose `apply` method was modified to accept a `ErrorList` too.

aleclearmind replied:
Some concerns has [here](https://github.com/revng/revng/pull/209#discussion_r1021809669).

drblallo replied:
this will disappear with no error lists

aleclearmind commented on tools/pipeline/print-invalidations/Main.cpp:
```diff
@@ -0,0 +1,88 @@
+  const auto &Ctx = Manager.context();
+  auto OriginalModel = *AbortOnError(Ctx.getGlobal<BinaryRef>(ModelGlobalName));
+
+  const auto &Name = ModelGlobalName;
+  auto AfterModel = AbortOnError(Ctx.getGlobal<BinaryRef>(Name));
```
Why are we model-specific? Can we make this tool non model-specific?

drblallo replied:
yeah i guess when we will have the need to do so we can change the signature of the tool, but there is not great need at the moment.

aleclearmind commented on include/revng/FunctionIsolation/IsolationFunctionKind.h:
```diff
@@ -0,0 +1,188 @@
+
+using yield::CrossRelations;
+
+// Find callers of a given `model::Function` via CrossRelations
```
Is there a real reason for functions below to be in a public header?

antoniofrighetto replied:
The idea was to expose those functions in a public header (but hide the inner implementation inside `detail` namespace) in order to make them available for `IsolatedFunctionKind.cpp` and for the upcoming `ABIEnforcedKind.cpp`.

aleclearmind replied:
This might be true for some of them, but not all. Please review what you have here.
For those you think will be useful in other invalidations, create a new header with an appropriate name.
Also consider having a `.cpp` file for implementations.

antoniofrighetto replied:
Sounds good: I added `InvalidationHelpers.h` under `Pipes` and moved those functions not needed to be exposed in the implementation file.

drblallo commented on tools/pipeline/Main.cpp:
```diff
@@ -178,12 +178,23 @@ int main(int argc, const char *argv[]) {
     auto Diff = AbortOnError(deserializeFileOrSTDIN<Type>(ApplyModelDiff));
+    revng::ErrorList EL;
+    Model.applyDiff(Diff, EL);
+    cantFail(std::move(EL));
```
not sure what you mean. It is just asserting that diffs do not fail

aleclearmind replied:
Gotcha. A couple of questions though:
1. Will it print the error messages within `ErrorList` nicely?
2. Is `cantFail` dropped in release?

drblallo replied:
1) i think it will print just the first error
2) no

antoniofrighetto committed:
Introduce helper for location conversion

antoniofrighetto commented on include/revng/Pipes/TupleTreeContainer.h:
```diff
@@ -0,0 +1,176 @@
+  std::unique_ptr<pipeline::ContainerBase>
+  cloneFiltered(const pipeline::TargetsList &Container) const final {
+    if (not Container.contains(pipeline::Target(*K)))
+      return std::make_unique<TupleTreeContainer>(this->name());
```
If we do not use `this` pointer to access `name` base class method, we need to accomplish it through direct invocation (`pipeline::ContainerBase::name()`). Isn't the first case preferred, here?

aleclearmind replied:
What stops you from just invoking `name()`?

```c++
struct X {
    void lel();
};

struct Y : public X {
   void mmm() { lel(); }
};
```

antoniofrighetto replied:
Right, however there must be some clash with names here, as I get:
```
/root/orchestra/sources/revng/include/revng/Pipes/TupleTreeContainer.h:94:51: error: use of undeclared identifier 'name'; did you mean 'LoadModelAnalysis::name'?
      return std::make_unique<TupleTreeContainer>(name());
                                                  ^~~~
                                                  LoadModelAnalysis::name
/root/orchestra/root/include/llvm/IR/PassManager.h:376:20: note: 'LoadModelAnalysis::name' declared here
  static StringRef name() {
                   ^
1 error generated.
```

antoniofrighetto commented on include/revng/TupleTree/TupleTreePath.h:
```diff
@@ -130,6 +135,21 @@ class ConcreteTupleTreeKeyWrapper : public TupleTreeKeyWrapper {
     }
   }
 
+  bool matches(const TupleTreeKeyWrapper &Other) const override {
```
It definitely is. Must have inadvertently manually taken it and not cherry-picked it this time. Fixed, apologize for the inconvenience.

antoniofrighetto commented on include/revng/Pipeline/Context.h:
```diff
@@ -103,7 +103,7 @@ class Context {
 
   bool containsReadOnlyContainer(llvm::StringRef Name) const {
     return ReadOnlyContainers.count(Name)
-           and ReadOnlyContainers.find(Name)->second != nullptr;
+           and ReadOnlyContainers.find(Name)->second->second != nullptr;
   }
```
Improved message for the bug fix.

antoniofrighetto commented on include/revng/Pipeline/Kind.h:
```diff
@@ -64,7 +65,9 @@ class Kind : public DynamicHierarchy<Kind> {
                                 pipeline::TargetsList &ToRemove,
-                                const GlobalTupleTreeDiff &Diff) const {}
+                                const GlobalTupleTreeDiff &Diff,
+                                const Global &Before,
+                                const Global &After) const {}
```
We use `After` (Model on which the diff has been applied) in `IsolatedKind::getInvalidations` to perform the difference between the current functions of the model and the old one, for FSO/PreservedRegisters.

antoniofrighetto commented on share/revng/pipelines/isolate-translate.yml:
```diff
@@ -2,6 +2,10 @@
 # This file is distributed under the MIT License. See LICENSE.md for details.
 #
 
+Containers:
+  - Name: cross-relations.yml
+    Type: BinaryCrossRelations
+    Role: CrossRelations
```
Closing this thread, as there's one already ongoing on the same concern.

aleclearmind commented on lib/FunctionIsolation/IsolatedFunctionKind.cpp:
```diff
@@ -0,0 +1,184 @@
+    // change).
+    for (const model::Function &Function : Model->Functions) {
+      MetaAddress Entry = Function.Entry;
+      if (OldModel->Functions.find(Entry) != OldModel->Functions.end())
+        if (haveFSOAndPreservedRegsChanged(OldModel->Functions.at(Entry)
```
Don't use `.at`, use `auto It = OldModel->Functions.find(Entry)`.

aleclearmind commented on lib/FunctionIsolation/IsolatedFunctionKind.cpp:
```diff
@@ -0,0 +1,184 @@
+      } else if (Change.Old) {
+        // Invalidate function's callers
+        auto Function = std::get<model::Function>(*Change.Old);
+        insertCallers(CrossRelations, Function.Entry, Targets);
```
Why not `insertCallersAndTransitiveClosureIfInline`?

antoniofrighetto replied:
As we have said, when removing a function, querying the `Relations` fields of `BinaryCrossRelations` should suffice in order to retrieve the callers, there should be no need to traverse the graph.

aleclearmind replied:
If a caller is inline, we need to recompute its callers too.

aleclearmind commented on lib/FunctionIsolation/IsolatedFunctionKind.cpp:
```diff
@@ -0,0 +1,184 @@
+      } else if (Change.Old) {
+        // Invalidate model::DynamicFunction's callers
+        auto DynamicFunction = std::get<model::DynamicFunction>(*Change.Old);
+        insertCallers(CrossRelations, DynamicFunction.OriginalName, Targets);
```
As above

aleclearmind commented on lib/FunctionIsolation/IsolatedFunctionKind.cpp:
```diff
@@ -0,0 +1,184 @@
+    auto MaybeFuncAttr = FuncAttrMatcher.match<MetaAddress>(Change.Path);
+    if (MaybeFuncAttr) {
+      auto Entry = std::get<MetaAddress>(*MaybeFuncAttr);
+      auto NewAttribute = std::get<FunctionAttribute>(*Change.New);
```
You're assuming attributes can only be removed, it seems.

antoniofrighetto replied:
You mean I'm assuming attributes can only be added?

aleclearmind replied:
Yep

aleclearmind commented on lib/FunctionIsolation/IsolatedFunctionKind.cpp:
```diff
@@ -0,0 +1,184 @@
+      // invalidate its callers (not the function itself as it would not provide
+      // any further information to the function itself).
+      if (NewAttribute == FunctionAttribute::NoReturn)
+        insertCallers(CrossRelations, Entry, Targets);
```
Why not `insertCallersAndTransitiveClosureIfInline`?

aleclearmind commented on include/revng/Pipes/InvalidationHelpers.h:
```diff
@@ -0,0 +1,152 @@
+                                          const TupleTree<model::Binary> &Model,
+                                          const std::string &Location,
+                                          std::set<MetaAddress> &Result) {
+  auto CRGraph = CR->toCallGraph();
```
You need to hoist this so you do it only once per call to `getInvalidation`.

antoniofrighetto replied:
Yeah, good catch, missed it.

aleclearmind commented on include/revng/Pipes/InvalidationHelpers.h:
```diff
@@ -0,0 +1,152 @@
+  auto VisitRange = llvm::inverse_post_order_ext(EntryNode, NoInlineFunctions);
+
+  for (const CRNode *Node : VisitRange) {
+    namespace p = pipeline;
```
Mmmmh, avoid this.

aleclearmind commented on include/revng/Pipes/InvalidationHelpers.h:
```diff
@@ -0,0 +1,152 @@
+
+  // Inverse DFS visit that aims at finding its callers, and its
+  // transitive closure, if they are all `Inline`.
+  llvm::df_iterator_default_set<const CRNode *> NoInlineFunctions;
+  auto VisitRange = llvm::inverse_post_order_ext(EntryNode, NoInlineFunctions);
+
+  for (const CRNode *Node : VisitRange) {
+    namespace p = pipeline;
+    namespace ranks = revng::ranks;
+
+    for (const CRNode *From : Node->predecessors()) {
+      auto MaybeKey = p::genericLocationFromString<0>(From->data(),
+                                                      ranks::Function,
+                                                      ranks::BasicBlock,
+                                                      ranks::Instruction);
+
+      if (MaybeKey.has_value()) {
+        auto &Function = Model->Functions.at(std::get<MetaAddress>(*MaybeKey));
+        if (Function.Attributes.count(model::FunctionAttribute::Inline) == 0)
+          NoInlineFunctions.insert(From);
+      }
+    }
+
+    auto Address = p::genericLocationFromString<0>(Node->data(),
+                                                   ranks::Function,
+                                                   ranks::BasicBlock,
+                                                   ranks::Instruction);
+    if (Address.has_value())
+      Result.insert(std::get<MetaAddress>(*Address));
+  }
+}
```
Can't you simply pre-populate `NoInlineFunctions` with all the functions that are not inline (except for the starting point)?

antoniofrighetto replied:
What's the point? It should be way more optimal to rely on `inverse_post_order_ext` to both traversal and filtering then first manually gathering non-inline functions (iterating over all the functions) and then do a conditional traversal. At least, that ought to be profiled to confirm which one of the two approaches is faster.

antoniofrighetto replied:
If `Inline` attribute is rare, it might be better to just gather a set of all the inline functions instead of listing the non-inline ones, and then the inner loop would just be a lookup on that set?

aleclearmind replied:
We're not discussing performance, we're discussing simplicity.
Using `_ext` without manually changing the set of visited nodes in the loop is much more intuitive and simpler to explain: "we mark all non-inline nodes as visited and then perform a simple inverse dfs".

If you're concerned about performance, you can reuse the initial set among various invocations of this method (copying it).

antoniofrighetto replied:
Following your direction, pre-populating `NoInlineFunctions` implies that, no sooner `inverse_post_order_ext ` later sees a node that is already in the set, it abandons that branch of the traversal, not exploring further. If the callers of the entry function are added to the set `NoInlineFunctions`, the inverse POT won't visit them. This means we end up getting neither the transitive closure nor the callers. To make it work, we would need to use the complementary set, namely, something like:
```diff
+  llvm::df_iterator_default_set<const CRNode *> AnotherSet;
-  auto VisitRange = llvm::inverse_post_order_ext(EntryNode, NoInlineFunctions);
+  auto VisitRange = llvm::inverse_post_order_ext(EntryNode, AnotherSet);

  for (const CRNode *Node : VisitRange) {
+    if (!llvm::is_contained(Node, NoInlineFunctions)) {
+      AnotherSet.emplace(Node);
+      continue;
+    }
    using namespace pipeline;
    namespace ranks = revng::ranks;

    auto Address = genericLocationFromString<0>(Node->data(),
                                                ranks::Function,
                                                ranks::BasicBlock,
                                                ranks::Instruction);
    if (Address.has_value())
      Result.insert(std::get<MetaAddress>(*Address));
  }
}
```
This seemingly adds further complexity. Personally, I would refrain from doing such a change.

aleclearmind replied:
It's more straightforward to read as opposed to "peeking" into successor's list while visiting nodes.
Also, you can pre-create once and for all the set of non-`Inline` functions and just copy it every time, so step 2) is not even there.

I already expressed these observations [in a previous comment](https://github.com/revng/revng/pull/209#discussion_r1029355688).

aleclearmind commented on share/revng/test/configuration/revng/for-invalidation.yml:
```diff
@@ -0,0 +1,98 @@
+      - type: revng.lifted
+        filter: for-invalidation
+    command: |-
+      BASELINE_LLVM_BITCODE="$$(mktemp)";
+      BASELINE_MODEL="$$(mktemp)";
+      BASELINE_MODEL_NAMED_DYN_FUNC="$$(mktemp)";
+      BASELINE_MODEL_OVERRIDEN="$$(mktemp)";
+      CFG="$$(mktemp)";
+      DIFF="$$(mktemp)";
+      TMP_FOLDER="$$(mktemp -d)";
+      trap 'rm -rf -- "$$BASELINE_LLVM_BITCODE" "$$BASELINE_MODEL" "$$DIFF" "$$TMP_FOLDER"' EXIT;
+      revng opt --collect-functions-from-callees --detect-abi --collect-cfg $INPUT  > "$$BASELINE_LLVM_BITCODE";
+      revng model dump < "$$BASELINE_LLVM_BITCODE" > "$$BASELINE_MODEL";
+      revng efa-extractcfg < "$$BASELINE_LLVM_BITCODE" > "$$CFG";
+      ./customname-and-inline-to-dynamicfunctions.py "$$CFG" "$$BASELINE_MODEL" > "$$BASELINE_MODEL_NAMED_DYN_FUNC";
+      revng model override-by-name "$$BASELINE_MODEL_NAMED_DYN_FUNC" ${SOURCE}.override.yml > "$$BASELINE_MODEL_OVERRIDEN";
+      revng pipeline
+        -i "$$BASELINE_LLVM_BITCODE":Lift/module.ll
+        -m "$$BASELINE_MODEL_OVERRIDEN"
+        --analyze Lift/DetectABI/module.ll/:Root
+        --resume "$$TMP_FOLDER"
+        --produce Isolate/module.ll/*:Isolated;
+      for FILE in ${SOURCE}.overrides/*.yml; do
+        revng model override-by-name "$$BASELINE_MODEL_OVERRIDEN" "$$FILE"
+          | ./remove-candidate-functions-from-model.py /dev/stdin "$$FILE"
+          | (revng model diff "$$BASELINE_MODEL_OVERRIDEN" /dev/stdin || true) > "$$DIFF";
+        revng print-invalidations --resume "$$TMP_FOLDER" "$$DIFF"
+          | ./metaaddress-to-customname.py "$$BASELINE_MODEL_OVERRIDEN" /dev/stdin
+          | diff -B /dev/stdin <(grep -vE '^\s*#' "$$FILE".invalidations);
+      done
```
Thoroughly comment what's going on, it's quite scary to look at as it is.

drblallo committed:
Remove ErrorList and introduce DiffError

drblallo committed:
Upcastable pointer upcast handle llvm::error

upcast was not able to handle rvalues nor was able to consume llvm
errors passed as ifNull argument. This commit fixes that issue.

aleclearmind commented on lib/Yield/Pipes/CallGraphPipes.cpp:
```diff
@@ -35,7 +35,10 @@ void ProcessCallGraph::run(pipeline::Context &Context,
+
+  // If some functions are missing, do not output anything
+  if (Metadata.size() != Model->Functions().size())
+    return;
```
Uhm, this doesn't seem right.

In order to produce the call graph we clearly need all of the functions or fail.

antoniofrighetto replied:
It's simply checking if the number of metadata entries we have is the same as the number of functions; if not – the call graph is likely to end up broken – we fail. I remember we were previously hard-asserting this, but some issue either with the pipeline or the build tests prevented us from doing it.

aleclearmind commented on include/revng/TupleTree/TupleTreeDiff.h:
```diff
@@ -514,6 +514,7 @@ TupleTreeDiff<T>::apply(TupleTree<T> &M, revng::ErrorList &EL) const {
       continue;
     }
     tupletreediff::detail::ApplyDiffVisitor<T> ADV{ &C, &EL };
+    M.evictCachedReferences();
```
I'm confused, why do this at every iteration of the loop? And why before declaring `ADV` and using it?

Can we do it once at the beginning? Or even better, not do it at all.

I think we could push `evictCachedReferences` into `initializeReferences`.

antoniofrighetto replied:
IIRC pushing `evictCachedReferences` exactly before `callByPath` was decided jointly with @pfez, though I actually don't see why it couldn't be hoisted, in principle. It still needs to be done before `callByPath`, I guess. Maybe @pfez can suggest an improvement here?

drblallo replied:
this can be closed

aleclearmind commented on lib/Yield/CrossRelations.cpp:
```diff
@@ -83,11 +83,8 @@ static void conversionHelper(const yield::CrossRelations &Input,
-      // TODO: handle dynamic function calls properly
       case yield::RelationType::IsDynamicallyCalledFrom:
+        AddEdge(LocationString, TargetString, RelationKind);
         break;
```
* No fixup commits. "TODO: handle dynamic function calls properly" was introduced in an earlier commit.
* Do we still have "handle dynamic function calls properly" or are they all handled?

> location-dependent `toCallGraph`

The commit message is a bit confusing: `s/location-dependent/location-based/`.

> some minor call graph specific improvements

Vague

antoniofrighetto replied:
Took responsibility for this, as some git shenanigans were required, and wanted to the relieve the poor @drblallo from this. So, the "TODO: handle dynamic function calls properly" fixup was fixed by squashing that change into the commit that introduced it (Ivan's one), plus some further fix down in the commits (to reflect this). The commit message has also been improved.

aleclearmind commented on include/revng/EarlyFunctionAnalysis/CallEdge.h:
```diff
@@ -46,15 +46,18 @@ class efa::CallEdge : public efa::generated::CallEdge {
-
-  CallEdge(MetaAddress Destination, FunctionEdgeType::Values Type) :
-    efa::generated::CallEdge(Destination, Type) {}
+  CallEdge(MetaAddress Destination, efa::FunctionEdgeType::Values Type) :
+    efa::generated::CallEdge() {
+    this->Destination() = Destination;
+    this->Type() = Type;
+  }
```
Why?
What's going on in this commit?

antoniofrighetto replied:
@drblallo, for some reason, possibly due to some commit in `develop`, EFA CallGraph & yield CrossRelation got desync'd. Sync'ing back is the purpose of this commit.

drblallo replied:
i can't offer a explanation better than antonion's one

aleclearmind commented on include/revng/Pipeline/Location.h:
```diff
@@ -176,6 +172,18 @@ locationFromString(const Rank &, std::string_view String) {
   return Location<Rank>::fromString(String);
 }
 
+/// A helper interface for location conversion.
+///
+/// It discloses the static `convert` member in an easier-to-access fashion.
+template<typename ResultRank, typename InputRank>
+  requires(RankConvertibleTo<ResultRank, InputRank>)
+inline constexpr Location<ResultRank>
+convertLocation(const ResultRank &Result, const Location<InputRank> &Input) {
+  return Location<ResultRank>::convert(Input);
+}
```
The argument is unused, this doesn't make much sense.

Cvelth replied:
(sorry for butting in into another person's PR, but)

It's exactly the same as the first argument of the function above it: this lets the caller to just pass `revng::ranks::Whatever` as an argument instead of trying to obtain its type (which is non-trivial) to pass as a template argument.

aleclearmind replied:
I'd prefer to use `decltype` in the caller then having an unused argument

Cvelth replied:
Are you sure you'd rather go
```cpp
auto NewLocation = pipeline::convertLocation<decltype(revng::ranks::Function)>(MyLocation);
```
over
```cpp
auto NewLocation = pipeline::convertLocation(revng::ranks::Function, MyLocation);
```

I'd say the second one is a lot easier to reason about.

aleclearmind replied:
Fair, it's a bit misleading for the user, but until he doesn't open the hood it's all good.

aleclearmind commented on tools/pipeline/print-invalidations/Main.cpp:
```diff
@@ -0,0 +1,67 @@
+  Diff.apply(AfterModel->get(), EL);
+  cantFail(std::move(EL));
+
+  TargetsList OverestimatedTargets;
```
Why "overestimated"?

drblallo replied:
because kinds do not know what is the content of containers, so they overexistamate the ones that will really be killed, and provide instead all that may be killed.

aleclearmind commented on share/revng/test/tests/invalidation/invalidation.c.override.yml:
```diff
@@ -0,0 +1,164 @@
+    CustomName: "prefix___libc_start_main"
+    Prototype: "/Types/CABIFunctionType-3475954398804502039"
+    Relocations:
+      - Address: "0x402000:Generic64"
```
* In tests we never ever commit hard-coded addresses.
* Drop all the stuff that you didn't write by hand such as `prefix___libc_start_main`

aleclearmind commented on share/revng/test/tests/invalidation/invalidation.c.override.yml:
```diff
@@ -0,0 +1,164 @@
+
+---
+Functions:
+  - CustomName: "bar"
+    OriginalName: bar
```
No need to have both.

aleclearmind commented on share/revng/test/tests/invalidation/invalidation.c.override.yml:
```diff
@@ -0,0 +1,164 @@
+
+---
+Functions:
+  - CustomName: "bar"
+    OriginalName: bar
+    Prototype: "/Types/RawFunctionType-13090241523659220176"
+  - CustomName: "foo"
+    OriginalName: foo
+    Prototype: "/Types/RawFunctionType-15357695279870196297"
+  - CustomName: "qux"
+    OriginalName: qux
+    Prototype: "/Types/RawFunctionType-6494433326490644482"
+  - CustomName: "baz"
+    OriginalName: baz
+    Prototype: "/Types/RawFunctionType-12005182500180626567"
```
I guess each function is trying to test something. Put meaningful names depending on what they test.

aleclearmind commented on share/revng/test/tests/invalidation/invalidation.c.overrides/CABIFunctionType-ABI.yml.invalidations:
```diff
@@ -0,0 +1,5 @@
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+plt___libc_start_main
```
Are you actually testing `plt___libc_start_main`?
It seems to me that you're committing whatever output you got, which is not how writing ground truth works.

aleclearmind commented:
1. Double check commit messages, are they indented at 72?
2. "FunctionIsolation: execute CollectCFG earlier": why? Improve explanation in the commit message.
3. Move "revng.daemon: accept no arguments for analysis" before "[Introduce ApplyDiffAnalysis](https://github.com/revng/revng/pull/209/commits/9bda21f6cb8cff4fe3806a63330c3b91656d7e80)"
4. Postpone "Allow access to Global in invalidation" before "Introduce IsolatedKind::getInvalidations"
5. The testing is not right, it seems that you have basically committed the output of the tool. This is not how testing works.
  You should 1) decide what to test (e.g., dropping a preserved register), 2) design the smallest possible test to test that only that, 3) write the ground truth.
  Also, you need to document what you're trying to test, I don't see that in the source file or in the ground truth. I have no idea how to check if the ground truth makes sense or not.
  If you're hardcoding addresses, or your ground truth contains names that you didn't come up with (e.g., `libc_start_main`) there is for sure something wrong. Tests should not depend on names in the libc or addresses/offsets that can easily change with minor configuration changes/version update of the compiler/linker.

aleclearmind commented:
1. Double check commit messages, are they indented at 72?
2. "FunctionIsolation: execute CollectCFG earlier": why? Improve explanation in the commit message.
3. Move "revng.daemon: accept no arguments for analysis" before "[Introduce ApplyDiffAnalysis](https://github.com/revng/revng/pull/209/commits/9bda21f6cb8cff4fe3806a63330c3b91656d7e80)"
4. Postpone "Allow access to Global in invalidation" before "Introduce IsolatedKind::getInvalidations"
5. The testing is not right, it seems that you have basically committed the output of the tool. This is not how testing works.
  You should 1) decide what to test (e.g., dropping a preserved register), 2) design the smallest possible test to test that only that, 3) write the ground truth.
  Also, you need to document what you're trying to test, I don't see that in the source file or in the ground truth. I have no idea how to check if the ground truth makes sense or not.
  If you're hardcoding addresses, or your ground truth contains names that you didn't come up with (e.g., `libc_start_main`) there is for sure something wrong. Tests should not depend on names in the libc or addresses/offsets that can easily change with minor configuration changes/version update of the compiler/linker.

aleclearmind commented:
The testing is not right, it seems that you have basically committed the output of the tool. This is not how testing works.
  You should 1) decide what to test (e.g., dropping a preserved register), 2) design the smallest possible test to test that only that, 3) write the ground truth.
  Also, you need to document what you're trying to test, I don't see that in the source file or in the ground truth. I have no idea how to check if the ground truth makes sense or not.
  If you're hardcoding addresses, or your ground truth contains names that you didn't come up with (e.g., `libc_start_main`) there is for sure something wrong. Tests should not depend on names in the libc or addresses/offsets that can easily change with minor configuration changes/version update of the compiler/linker.

drblallo commented on lib/Pipeline/Runner.cpp:
```diff
@@ -395,21 +395,27 @@ void Runner::getDiffInvalidations(const GlobalTupleTreeDiff &Diff,
                                   const Global &Before,
                                   const Global &After,
                                   InvalidationMap &Map) const {
+  TargetsList OverestimatedTargets;
```
done, for some reason i can't mark this as resolved

antoniofrighetto committed:
`override-by-name.py`: introduce `--remove-others`

drblallo committed:
Rework yield and efa cross relations

Yield and efa shared the same namespace, this commits splits them and
introduces tuple tree containers as well.

drblallo commented on include/revng/FunctionIsolation/IsolationFunctionKind.h:
```diff
@@ -0,0 +1,188 @@
+inline TaggedFunctionKind
+  ABIEnforced("ABIEnforced", &ranks::Function, FunctionTags::ABIEnforced);
+
+using yield::CrossRelations;
```
this issue can be closed
